### PR TITLE
Add property conn_name and connection to dbapi_hook

### PR DESCRIPTION
**Pull request by zhongjiajie**
---
Issue link: WILL BE INSERTED BY [boring-cyborg](https://github.com/kaxil/boring-cyborg)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by zhongjiajie** [comment]
We have one subclass hook still WIP in https://github.com/apache/airflow/pull/7901

File: airflow/hooks/dbapi_hook.py
Comment (zhongjiajie) [inline airflow/hooks/dbapi_hook.py]: I change this un related cause it take me a little time to get the point, I think we should use `if condition ... else if ... else <default_value>` for more straightforward.

File: airflow/hooks/dbapi_hook.py
Comment (feluelle) [inline airflow/hooks/dbapi_hook.py]: We don't add production code only because a linter is unhappy!

What it give us is that it..
```suggestion
        Get the name of the database connection attribute which identifies the connection.
```

File: airflow/hooks/dbapi_hook.py
Comment (feluelle) [reply airflow/hooks/dbapi_hook.py]:
```python
else:
    setattr(self, self.conn_name_attr, kwargs.get(self.conn_name_attr, self.default_conn_name)
```
instead of
```python
elif self.conn_name_attr in kwargs:
    setattr(self, self.conn_name_attr, kwargs[self.conn_name_attr])
else:
    setattr(self, self.conn_name_attr, self.default_conn_name)
```
WDYT?

Comment (zhongjiajie) [reply airflow/hooks/dbapi_hook.py]: Oh, that much simple, but I have half-memories use `dict[key]` for exactly key as best practices. WDYT.

File: airflow/hooks/dbapi_hook.py
Comment (zhongjiajie) [reply airflow/hooks/dbapi_hook.py]: You're right

File: airflow/hooks/dbapi_hook.py
Comment (zhongjiajie) [reply airflow/hooks/dbapi_hook.py]: Oh, that much simple, but I have half-memories use `dict[key]` for exactly key as best practices. WDYT.

File: airflow/hooks/dbapi_hook.py
Comment (zhongjiajie) [reply airflow/hooks/dbapi_hook.py]: You're right.

File: airflow/hooks/dbapi_hook.py
Comment (zhongjiajie) [reply airflow/hooks/dbapi_hook.py]: Add it

**Comment by zhongjiajie** [comment]
For reviewer, in https://github.com/apache/airflow/pull/7903/commits/cf0a5f6fca8f219d0038c4c7ace5d98695400ce4 I just add property `conn_name` to get `conn_name`. And then I find out we could change `conn = self.get_connection(self.conn_name)` to `conn = self.get_connection()` due to `self.conn_name` almost same in subclass of `DBAPI` and is no need, So I add function `get_connection` in `dbapi_hook` to get conn_name if no conn_id pass from subclass.

**Review by potiuk — COMMENTED** [review]
It looks good. I would like to understand though why it was done like that in the first place. seems unnecessary complex - maybe others (@feluelle @mik-laj can take a look and shed some light on it ) :)

File: airflow/hooks/dbapi_hook.py
Comment (feluelle) [inline airflow/hooks/dbapi_hook.py]:
```suggestion
        conn_name = conn_id or self.conn_name
```

File: airflow/hooks/dbapi_hook.py
Comment (feluelle) [reply airflow/hooks/dbapi_hook.py]: You missed that.

File: airflow/hooks/dbapi_hook.py
Comment (feluelle) [inline airflow/hooks/dbapi_hook.py]:
```suggestion
        Get the database connection object.
```

**Comment by feluelle** [comment]
> It looks good. I would like to understand though why it was done like that in the first place. seems unnecessary complex - maybe others (@feluelle @mik-laj can take a look and shed some light on it ) :)

I agree with you Jarek. I also don't know why it is as it is. 🤔

File: airflow/hooks/dbapi_hook.py
Comment (zhongjiajie) [reply airflow/hooks/dbapi_hook.py]: Sorry, I just force push and forget fetch from my origin

**Comment by zhongjiajie** [comment]
My laptop almost power off and I forget the charger in office, have to continue in tomorrow.

File: airflow/hooks/dbapi_hook.py
Comment (zhongjiajie) [inline airflow/hooks/dbapi_hook.py]: Emmy, I think we get some problem here, `mypy` unhappy and said `76: error: Signature of "get_connection" incompatible with supertype "BaseHook"`, because in `BaseHook` function `get_connection` is `classmethod`, but in `DbApiHook` function `get_connection` we use `getattr(self, self.conn_name_attr)` so have to be instance method.

And `DbApiHook` we do not need to pass `conn_id` anymore, and I think we should change function `get_connection` to `@property connection` here. WDYT

Comment (zhongjiajie) [reply airflow/hooks/dbapi_hook.py]: Or maybe we could just use `# type: ignore` to ignore `mypy` in L76

**Comment by zhongjiajie** [comment]
And for now, the solution two pass CI, and I think it work.

The question is, should we take solution one and change `get_connection` in `dbapi_hook` to property `connection`? If we use solution one, subclass of `dbapi_hook` have to use `self.connection` to get connection which is different from subclass from `base_hook`(using `get_connection`)

But if we take solution two, `get_connection` don't pass `mypy` check, and `self.get_connection` and `DbApiHook.get_connection` call different method, I think it will confuse user when calling it

**Comment by zhongjiajie** [comment]
Two solutions CI passed, we have **two fix solutions** here, I prefer **the second one**, and it also the PR final status

* Add `# type: ignore` in L76  in dbapi_hook to avoid `error: Signature of "get_connection" incompatible with supertype "BaseHook"`
```diff
-    def get_connection(self, conn_id: Optional[str] = None) -> Connection:
+    def get_connection(self, conn_id: Optional[str] = None) -> Connection:  # type: ignore
```
* Using property function `connection` in dbapi_hook to get connection, which complete fix Signature error, and make dbapi_hook more easier.

File: airflow/hooks/dbapi_hook.py
Comment (feluelle) [inline airflow/hooks/dbapi_hook.py]: I would prefer going for the first option and don't add another `connection` property. And use `self.get_connection(self.conn_name)` in hooks.

Comment (feluelle) [inline airflow/hooks/dbapi_hook.py]:
```suggestion
        return self.get_connection(self.conn_name)
```
should also work, doesn't it?

Actually I think this option (the second one) is better. 👍

Comment (zhongjiajie) [reply airflow/hooks/dbapi_hook.py]: Haha, I think so, and I open this to hear other committer idea

**Comment by zhongjiajie** [comment]
Reopen this stale PR

**Comment by zhongjiajie** [comment]
@feluelle @potiuk @kaxil @kaxil @turbaszek I think I need one approving review here. I checked and find out there no `DbApiHook` subclass add during this time, which mean this patch cover/change all `DbApiHook` subclass.

**Review by turbaszek — APPROVED** [review]

**Comment by zhongjiajie** [comment]
Ok, we have +1 here, if no one disapproval here I will merge this patch to master.

**Comment by zhongjiajie** [comment]
Ok, we have +1 here, if no one disapproval here I will merge this patch to master.

**Comment by feluelle** [comment]
Please rebase once again - just to be sure. :)

**Comment by zhongjiajie** [comment]
Thanks @turbaszek @feluelle, now I rebase and push to restart the CI, will merge when it pass

File: airflow/providers/odbc/hooks/odbc.py
Comment (feluelle) [inline airflow/providers/odbc/hooks/odbc.py]:
```suggestion
    def connection(self):
```
You can overwrite it.

File: airflow/providers/odbc/hooks/odbc.py
Comment (feluelle) [inline airflow/providers/odbc/hooks/odbc.py]:
```suggestion
            self._connection = super().connection
```

File: airflow/providers/snowflake/hooks/snowflake.py
Comment (feluelle) [inline airflow/providers/snowflake/hooks/snowflake.py]: This is unrelated.

File: tests/providers/mysql/hooks/test_mysql.py
Comment (feluelle) [inline tests/providers/mysql/hooks/test_mysql.py]:
```suggestion
        self.db_hook.connection = mock.Mock(return_value=self.connection)
```

File: tests/providers/mysql/hooks/test_mysql.py
Comment (feluelle) [inline tests/providers/mysql/hooks/test_mysql.py]:
```suggestion
        self.db_hook.connection = mock.Mock(return_value=self.connection)
```

File: tests/providers/odbc/hooks/test_odbc.py
Comment (feluelle) [inline tests/providers/odbc/hooks/test_odbc.py]: Please use the `Connection` class to init a connection like in the other cases.

For example:
```python
self.conn = mock.Mock(return_value=Connection(...))
```

**Review by feluelle — CHANGES_REQUESTED** [review]
Actually I am a bit confused by all these `connection`, `_connection`, ... parameters.

Wouldn't it be easier if we just use the `conn_id` like we do in other hooks which are not based on dbapi_hook ?

**Comment by feluelle** [comment]
So my suggestion is to remove `conn_name_attr` and add a `*_conn_id` parameter to init. WDYT @turbaszek @zhongjiajie ?

As it is now I am not a fan of it anymore I am sorry @zhongjiajie but this looks really too complex, don't you think?

**Comment by zhongjiajie** [comment]
I think we should keep `conn_name_attr` but change property name which confused users, like `connection` as you said above. Cause we need dbapi_hook as parent class to handle some RDBMS like hook.

File: airflow/providers/snowflake/hooks/snowflake.py
Comment (zhongjiajie) [reply airflow/providers/snowflake/hooks/snowflake.py]: I have to change it to pass CI, I have no idea why previous code could pass the CI, mypy will raise error cause `aws_access_key_id, aws_secret_access_key` maybe not define

**Comment by ashb** [comment]
@zhongjiajie Are you going to carry on with this PR? Where did we get to with it when we left off?

**Pull request closed without merging**
